wrong type deduction inside the compile time for loop

### Describe the bug

inside the generated c-file, at each iteration of the loop the type of myvar is MyStruct instead of the underlying type returned by the function.

### Expected Behavior

program output:
```
int
string
bool
```

### Current Behavior

```
/tmp/v_1000/main.4213678605308598159.tmp.c:12226: warning: implicit declaration of function 'main__get_type_T_int'
/tmp/v_1000/main.4213678605308598159.tmp.c:12226: error: '{' expected (got ";")
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```


### Reproduction Steps

```V
struct MyStruct {
	a int
	b string
	c bool
}

fn get_type[T](t T) T {
	return t
}

fn main() {
    my := MyStruct{}
	$for field in MyStruct.fields {
		var := get_type(my.$(field.name))
		println(typeof(var).name)
	}
}
```


### Possible Solution

_No response_

### Additional Information/Context

/tmp/v_1000/main.4213678605308598159.tmp.c:12226:
```C
void main__main(void) {
        main__MyStruct my = ((main__MyStruct){.a = 0,.b = (string){.str=(byteptr)"", .is_lit=1},.c = 0,});
        /* $for field in main.MyStruct(fields) */ {
                        FieldData field = {0};
                /* field 0 */ {
                        field.name = _SLIT("a");
                        field.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
                        field.typ = 7;
                        field.unaliased_typ = 7;
                        field.is_pub = false;
                        field.is_mut = false;
                        field.is_shared = false;
                        main__MyStruct var = main__get_type_T_int(my.a);
                        println(_SLIT("MyStruct"));
                }
                /* field 1 */ {
                        field.name = _SLIT("b");
                        field.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
                        field.typ = 20;
                        field.unaliased_typ = 20;
                        field.is_pub = false;
                        field.is_mut = false;
                        field.is_shared = false;
                        main__MyStruct var = main__get_type_T_string(my.b);
                        println(_SLIT("MyStruct"));
                }
                /* field 2 */ {
                        field.name = _SLIT("c");
                        field.attrs = __new_array_with_default(0, 0, sizeof(string), 0);
                        field.typ = 18;
                        field.unaliased_typ = 18;
                        field.is_pub = false;
                        field.is_mut = false;
                        field.is_shared = false;
                        main__MyStruct var = main__get_type_T_bool(my.c);
                        println(_SLIT("MyStruct"));
                }
        }// $for
}
```

### V version

V 0.3.2 0128d2d.6a179a2

### Environment details (OS name and version, etc.)

```
OS: linux, "Arch Linux"
Processor: 2 cpus, 64bit, little endian, Intel(R) Celeron(R) CPU B820 @ 1.70GHz
CC version: cc (GCC) 12.2.0

getwd: /home/anotherusuario
vmodules: /home/anotherusuario/.vmodules
vroot: /usr/lib/vlang
vexe: /usr/lib/vlang/v
vexe mtime: 2022-12-24 04:35:51
is vroot writable: false
is vmodules writable: true
V full version: V 0.3.2 0128d2d.6a179a2

Git version: git version 2.39.0
Git vroot status: Error: fatal: not a git repository (or any of the parent directories): .git
.git/config present: false
thirdparty/tcc status: Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc
 Error: fatal: detected dubious ownership in repository at '/usr/lib/vlang/thirdparty/tcc'
To add an exception for this directory, call:

	git config --global --add safe.directory /usr/lib/vlang/thirdparty/tcc
```

--- a/vlib/v/ast/ast.v
+++ b/vlib/v/ast/ast.v
@@ -590,6 +590,7 @@ pub mut:
 	is_ctor_new        bool // if JS ctor calls requires `new` before call, marked as `[use_new]` in V
 	args               []CallArg
 	expected_arg_types []Type
+	comptime_ret_val   bool
 	language           Language
 	or_block           OrExpr
 	left               Expr // `user` in `user.register()`

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -871,7 +871,26 @@ fn (mut c Checker) infer_fn_generic_types(func ast.Fn, mut node ast.CallExpr) {
 					func_.name = ''
 					idx := c.table.find_or_register_fn_type(func_, true, false)
 					typ = ast.new_type(idx).derive(arg.typ)
+				} else if c.inside_comptime_for_field && sym.info is ast.Struct
+					&& arg.expr is ast.ComptimeSelector {
+					compselector := arg.expr as ast.ComptimeSelector
+					if compselector.field_expr is ast.SelectorExpr {
+						selectorexpr := compselector.field_expr as ast.SelectorExpr
+						if selectorexpr.expr is ast.Ident {
+							ident := selectorexpr.expr as ast.Ident
+							if ident.name == c.comptime_for_field_var {
+								typ = c.comptime_fields_default_type
+
+								if func.return_type.has_flag(.generic)
+									&& gt_name == c.table.type_to_str(func.return_type) {
+									c.error('cannot use value known only on compile-time as return',
+										node.pos)
+								}
+							}
+						}
+					}
 				}
+
 				if arg.expr.is_auto_deref_var() {
 					typ = typ.deref()
 				}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -883,8 +883,7 @@ fn (mut c Checker) infer_fn_generic_types(func ast.Fn, mut node ast.CallExpr) {
 
 								if func.return_type.has_flag(.generic)
 									&& gt_name == c.table.type_to_str(func.return_type) {
-									c.error('cannot use value known only on compile-time as return',
-										node.pos)
+									node.comptime_ret_val = true
 								}
 							}
 						}

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -95,6 +95,7 @@ mut:
 	vweb_gen_types                   []ast.Type // vweb route checks
 	timers                           &util.Timers = util.get_timers()
 	for_in_any_val_type              ast.Type
+	comptime_for_field_var           string
 	comptime_fields_default_type     ast.Type
 	comptime_fields_type             map[string]ast.Type
 	fn_scope                         &ast.Scope = unsafe { nil }

--- a/vlib/v/checker/comptime.v
+++ b/vlib/v/checker/comptime.v
@@ -180,6 +180,7 @@ fn (mut c Checker) comptime_for(node ast.ComptimeFor) {
 			sym_info := sym.info as ast.Struct
 			c.inside_comptime_for_field = true
 			for field in sym_info.fields {
+				c.comptime_for_field_var = node.val_var
 				c.comptime_fields_type[node.val_var] = node.typ
 				c.comptime_fields_default_type = field.typ
 				c.stmts(node.stmts)

--- a/vlib/v/checker/comptime.v
+++ b/vlib/v/checker/comptime.v
@@ -189,6 +189,7 @@ fn (mut c Checker) comptime_for(node ast.ComptimeFor) {
 				tsym := c.table.sym(unwrapped_expr_type)
 				c.table.dumps[int(unwrapped_expr_type.clear_flag(.optional).clear_flag(.result))] = tsym.cname
 			}
+			c.comptime_for_field_var = ''
 			c.inside_comptime_for_field = false
 		}
 	} else {

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -136,7 +136,7 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				is_auto_heap = left.obj.is_auto_heap
 			}
 		}
-		styp := g.typ(var_type)
+		mut styp := g.typ(var_type)
 		mut is_fixed_array_init := false
 		mut has_val := false
 		match val {
@@ -146,7 +146,12 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 			}
 			ast.CallExpr {
 				is_call = true
-				return_type = val.return_type
+				if val.comptime_ret_val {
+					return_type = g.comptime_for_field_type
+					styp = g.typ(return_type)
+				} else {
+					return_type = val.return_type
+				}
 			}
 			// TODO: no buffer fiddling
 			ast.AnonFn {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1234,6 +1234,7 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 	mut is_interface_call := false
 	mut is_selector_call := false
 	mut has_comptime_field := false
+	mut comptime_args := []int{}
 	if node.left_type != 0 {
 		left_sym := g.table.sym(node.left_type)
 		if left_sym.kind == .interface_ {
@@ -1266,10 +1267,12 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 					node_.args[i].typ = call_arg.expr.obj.typ
 					if call_arg.expr.obj.is_comptime_field {
 						has_comptime_field = true
+						comptime_args << i
 					}
 				}
 			} else if mut call_arg.expr is ast.ComptimeSelector {
 				has_comptime_field = true
+				comptime_args << i
 			}
 		}
 	}
@@ -1362,7 +1365,11 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 			if func.generic_names.len > 0 {
 				if g.comptime_for_field_type != 0 && g.inside_comptime_for_field
 					&& has_comptime_field {
-					name = g.generic_fn_name([g.comptime_for_field_type], name)
+					mut concrete_types := node.concrete_types.map(g.unwrap_generic(it))
+					for k in comptime_args {
+						concrete_types[k] = g.comptime_for_field_type
+					}
+					name = g.generic_fn_name(concrete_types, name)
 				} else {
 					concrete_types := node.concrete_types.map(g.unwrap_generic(it))
 					name = g.generic_fn_name(concrete_types, name)

--- a/vlib/v/tests/comptime_on_generics_func_test.v
+++ b/vlib/v/tests/comptime_on_generics_func_test.v
@@ -1,0 +1,25 @@
+module main
+
+struct MyStruct {
+	b string
+	c bool
+}
+
+fn get_type[T](t T) T {
+	return t
+}
+
+
+fn test_main() {
+    my := MyStruct{'cool', false}
+	$for field2 in MyStruct.fields {
+		$if field2.typ is string {
+			var := get_type(my.$(field2.name))
+			assert dump(var) == 'cool'
+		}
+		$if field2.typ is bool {
+			var := get_type(my.$(field2.name))
+			assert dump(var).str() == 'false'
+		}
+	}
+}

--- a/vlib/v/tests/comptime_on_generics_func_test.v
+++ b/vlib/v/tests/comptime_on_generics_func_test.v
@@ -9,9 +9,8 @@ fn get_type[T](t T) T {
 	return t
 }
 
-
 fn test_main() {
-    my := MyStruct{'cool', false}
+	my := MyStruct{'cool', false}
 	$for field2 in MyStruct.fields {
 		$if field2.typ is string {
 			var := get_type(my.$(field2.name))

--- a/vlib/v/tests/comptime_var_on_multiple_args_test.v
+++ b/vlib/v/tests/comptime_var_on_multiple_args_test.v
@@ -1,0 +1,33 @@
+module main
+
+fn ok3[T](o T) T {
+	return o
+}
+
+fn ok2[T](o T, s string) string {
+	return s
+}
+
+fn ok[T](o T, s string) string {
+	return "${o}${s}"
+}
+
+struct Test {
+	s string
+}
+
+fn test_main() {
+	s := Test{'foo'}
+
+	$for f in Test.fields {
+		assert ok3(s.$(f.name)).str() == 'foo'
+	}
+
+	$for f in Test.fields {
+		assert ok(s.$(f.name), 'bar') == 'foobar'
+	}
+
+	$for f in Test.fields {
+		assert ok2(s.$(f.name), 'ok') == 'ok'
+	}
+}

--- a/vlib/v/tests/comptime_var_on_multiple_args_test.v
+++ b/vlib/v/tests/comptime_var_on_multiple_args_test.v
@@ -9,7 +9,7 @@ fn ok2[T](o T, s string) string {
 }
 
 fn ok[T](o T, s string) string {
-	return "${o}${s}"
+	return '${o}${s}'
 }
 
 struct Test {
@@ -22,11 +22,9 @@ fn test_main() {
 	$for f in Test.fields {
 		assert ok3(s.$(f.name)).str() == 'foo'
 	}
-
 	$for f in Test.fields {
 		assert ok(s.$(f.name), 'bar') == 'foobar'
 	}
-
 	$for f in Test.fields {
 		assert ok2(s.$(f.name), 'ok') == 'ok'
 	}

